### DISCUSSION TITLE
Collect DC current/voltage information

### DISCUSSION THREAD

**Discussion by amyeguo** [microsoft/jacdac#1208] [General]
Hello! I’m an undergraduate student in the Ka Moamoa lab at Northwestern with Josiah Hester, I’m interested in extending JACDAC for the micro:bit to help developers visually debug and understand the energy usage of their programs.

I’m currently trying to import a forked copy of pxt-jacdac as an extension into MakeCode. But when I try to “Add Blocks” for the button2 it runs into this error:
<img width="1781" alt="Screen Shot 2022-08-30 at 11 26 40 AM" src="https://user-images.githubusercontent.com/72362221/187492631-d52f4d0e-f1c1-445e-816e-9284633b7e06.png">

I was wondering if there are any missing steps besides cloning & importing the URL for a forked copy of pxt-jacdac to get the JACDAC extension to work.

Thanks!

**Comment by pelikhan** [comment]
we typically don't use the online editor to work on the pxt-jacdac project (and sub-projects). instead follow the developer instructions in https://github.com/microsoft/pxt-jacdac#developer-section

May I ask what you are trying to acheive? pxt-jacdac is the core library and it might be easier for you to work by simply referencing it.

**Reply by amyeguo** [reply]
Apologies! I think my initial post was unclear, I was interested in extending the Jacdac web tools to collect the device’s energy usage over a period of time, maybe similar to https://microsoft.github.io/jacdac-docs/tools/collector/

Do you think this is feasible using this Jacdac Javascript package? https://microsoft.github.io/jacdac-docs/clients/javascript/

**Reply by pelikhan** [reply]
let me add a sample that gets you closer to that. How do you plan to read the power usage? are you relyaing on the https://microsoft.github.io/jacdac-docs/services/power/ or do you plan to create a new service for this?

**Reply by pelikhan** [reply]
it's definitely possible. do you want to run this in a web page (probably the easiest) or as a node.js script?

**Reply by amyeguo** [reply]
I think web page to start. For the pure power provider devices, I think we were planning to use the power service you linked. For the pure power consumer devices, I was looking at the [DC Current Measurement](https://microsoft.github.io/jacdac-docs/services/dccurrentmeasurement/) and [DC Voltage Measurement](https://microsoft.github.io/jacdac-docs/services/dcvoltagemeasurement/) services.

**Reply by pelikhan** [reply]
here's a sample that listens for dc current measurement services

https://codesandbox.io/s/vanilla-jacdac-typescript-forked-pgwhwc?file=/index.ts

The interresting bit is to grab dc services when they get announce and register for each "report_receive" events,.

```
// listen for new device
bus.on(DEVICE_ANNOUNCE, (device: JDDevice) => {
  log(device.describe());
  // go through all dc_current_measurement services
  const services = device.services({
    serviceClass: SRV_DC_CURRENT_MEASUREMENT
  });
  services.forEach((service) => {
    // grab reading register
    const reading = service.readingRegister;
    // register handler for each receive data entry
    reading.on(REPORT_RECEIVE, () => {
      // read data
      const dc = reading.unpackedValue[0] as number;
      log(`${device}: ${dc}A`);
    });
  });
});
```

**Reply by pelikhan** [reply]
there is no simulator support for DC current measurement yet but i could add that

**Reply by pelikhan** [reply]
I have added support in the simulator for a DC service. 
![image](https://user-images.githubusercontent.com/4175913/188171471-1109423c-7bc6-4952-b77c-0a1427c0eb22.png)

![image](https://user-images.githubusercontent.com/4175913/188173421-c7c4be5f-8afc-42e8-965f-067f44b7542e.png)

**Reply by amyeguo** [reply]
Thank you so much! If we were to create a power provider as a JacDac device such as a solar panel, could we profile the DC current/voltage measurement of that device?

**Reply by pelikhan** [reply]
Once the volutage/current is availabe, you coudl build all kinds of tools around the jacdac client SDK. Including profilers.
